vermouth/pdb/pdb.py

```diff
@@ -16,13 +16,408 @@
+            ('resname', str, 4),
+            ('chain', str, 1),
+            ('resid', int, 4),
+            ('insertion_code', str, 1),
+            ('', str, 3),
+            ('x', float, 8),
+            ('y', float, 8),
+            ('z', float, 8),
+            ('occupancy', float, 6),
+            ('temp_factor', float, 6),
+            ('', str, 10),
+            ('element', str, 2),
+            ('charge', str, 2),
+        ]
+
+        start = 0
+        field_slices = []
+        for name, type_, width in fields:
+            if name:
+                field_slices.append((name, type_, slice(start, start + width)))
+            start += width
+
+        properties = {}
+        for name, type_, slice_ in field_slices:
+            properties[name] = type_(line[slice_].strip())
+
+        pos = (properties.pop('x'), properties.pop('y'), properties.pop('z'))
+        # Coordinates are read in Angstrom, but we want them in nm
+        properties['position'] = np.array(pos, dtype=float) / 10
+
+        if not properties['element']:
+            atomname = properties['atomname']
+            properties['element'] = first_alpha(atomname)
+        if properties['altloc'] not in ['', 'A']:
+            # TODO: allow selecting alternative conformation for specific
+            #       residues.
+            LOGGER.warning("There is an alternative conformation for atom {}. "
+                           "We use conformation A exclusively",
+                           format_atom_string(properties),
+                           type='inconsistent-data')
+            return
+        if (properties['resname'] in self.exclude or
+                (self.ignh and properties['element'] == 'H')):
+            return
+        idx = max(self.active_molecule) + 1 if self.active_molecule else 0
+        self.active_molecule.add_node(idx, **properties)
+
+    atom = _atom
+    hetatm = _atom
+
+    def model(self, line, lineno=0):
+        """
+        Parse a MODEL record. If the model is not the same as :attr:`modelidx`,
+        this model will not be parsed.
+
+        Parameters
+        ----------
+        line: str
+            The line to parse. Should start with "MODEL ", but this is not
+            checked.
+        lineno: int
+            The line number (not used).
+        """
+        try:
+            modelnr = int(line[10:14])
+        except ValueError:
+            return
+        else:
+            self._skipahead = modelnr != self.modelidx
+
+    def conect(self, line, lineno=0):
+        """
+        Parse a CONECT record. The line is stored for later processing.
+
+        Parameters
+        ----------
+        line: str
+            The line to parse. Should start with CONECT, but this is not checked
+        lineno: int
+            The line number (not used).
+        """
+        # We can't add edges immediately, since the molecule might not be parsed
+        # yet (does the PDB file format mandate anything on the order of
```

Comment:
CONECT records are always at the end of the file. The PDB format specification gives an order of records: http://www.wwpdb.org/documentation/file-format-content/format33/sect1.html
If this may lead to simplification of the code, then it could be listed as an issue.

Reply:
It would simplify the code, but this makes us a bit more resilient to badly formatted pdb files.

Reply:
I concur.